### PR TITLE
chore(metrics): emit sign_up events from GA4 client side

### PR DESCRIPTION
**Pull request by tianrunhe**
# Description

Today, we sent `sign_up` events from Segment to GA4. However, due to limitation of Segment, these events have no session information (or any info):

Screenshot 2023-01-09 at 3 56 12 PM

As a result, `sign_up` events cannot be attributed to previous `page_view` events in GA4:

211431799-ecd7bbd7-eef0-4a23-89a2-0353f06454ee

Screenshot 2023-01-09 at 4 01 24 PM

Basically no useful analysis can be performed...😢

This PR emits `sign_up` events directly from GA4 client side. This way all relevant info (session, geo, ...) are automatically populated.

## Demo

The GA4 client side emitted `sign_up` events look like these in BigQuery:

Screenshot 2023-01-09 at 4 06 29 PM

Attribution seems to work afterwards:

Screenshot 2023-01-09 at 4 08 06 PM

Screenshot 2023-01-09 at 4 07 26 PM

## Testing scenarios

- [ ] Sign up with username and password
  - Verify a `sign_up` event is emitted to GA4

- [ ] Sign up with Google
  - Verify a `sign_up` event is emitted to GA4

- [ ] Log in with Google
  - Verify **NO** `sign_up` event is emitted to GA4

## Final checklist

- [x] I checked the [code review guidelines](../docs/codeReview.md)
- [x] I have added [Metrics Representative](../docs/codeReview.md#metrics-representative) as reviewer(s) if my PR invovles metrics/data/analytics related changes
- [x] I have performed a self-review of my code, the same way I'd do it for any other team member
- [x] I have tested all cases I listed in the testing scenarios and I haven't found any issues or regressions
- [x] Whenever I took a non-obvious choice I added a comment explaining why I did it this way
- [x] I added the label `One Review Required` if the PR introduces only minor changes, does not contain any architectural changes or does not introduce any new patterns and I think one review is sufficient'
- [x] PR title is human readable and could be used in changelog

### CODE REVIEW

File: packages/client/components/AnalyticsPage.tsx [hunk 1]
```diff
@@ -89,39 +89,39 @@ const TIME_TO_RENDER_TREE = 100
 
 const AnalyticsPage = () => {
   const atmosphere = useAtmosphere()
-  const [isGAInitialized, setIsGAInitialized] = useState(false)
+  if (gaMeasurementId) {
+    ReactGA.initialize(gaMeasurementId, {
```
Comment (tianrunhe) [inline packages/client/components/AnalyticsPage.tsx:93]: There won't be duplicate initializations as it's [protected](https://github.com/PriceRunner/react-ga4/blob/master/src/ga4.js#L183) inside `ReactGA`

File: packages/client/components/AnalyticsPage.tsx [hunk 2]
```diff
@@ -89,39 +89,39 @@ const TIME_TO_RENDER_TREE = 100
...
+        const {viewer} = res
+        const {id, segmentId} = viewer
+        ReactGA.set({
+          userId: id,
+          clientId: segmentId ?? getAnonymousId()
         })
+      } else {
+        ReactGA.set({
+          userId: null,
+          clientId: getAnonymousId()
```
Comment (tianrunhe) [inline packages/client/components/AnalyticsPage.tsx:119]: We need the `clientId` before login & after logout to attribute the entire user journey

**Review by Dschoordsch — APPROVED** [review]
I had some trouble testing this, after logging out it did not reliably send any events anymore, even with a new user. I trust you that you've tested this properly and this is a local issue.

**Comment by Dschoordsch** [comment]
Also, is it intentional that we don't track SAML signups?

**Comment by tianrunhe** [comment]
> Also, is it intentional that we don't track SAML signups?

I completely overlooked this use case. Thanks for pointing it out! What do you think my implementation here? It's a bit different than sign up with password or Google.

**Review by Dschoordsch — COMMENTED** [review]
Code looks good, I just haven't come around setting up SAML for testing yet, will do tomorrow.

**Review by Dschoordsch — APPROVED** [review]
Debugging GA4 still feels a bit like a gamble, but the SAML event works just fine, good work!

**Pull request merged**